### PR TITLE
Add `context` awareness in services and `cmd`s. Closes #1206. Closes #1214. Closes #1215

### LINKED ISSUE #1206: Introduce `context` awareness to service installation and init

### LINKED ISSUE #1214: `DbClient.GetSchemaFromDB` should operate under a `context`

### LINKED ISSUE #1215: Database connection establishment should work under a `context`
`db/db_client/db_client.go` `NewDbClient` should take in the parent `ctx`

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: cmd/service.go [hunk 1]
```diff
@@ -186,9 +190,9 @@ func runServiceStartCmd(cmd *cobra.Command, args []string) {
 		}
 	}
 
-	err = db_local.RefreshConnectionAndSearchPaths(invoker)
+	err = db_local.RefreshConnectionAndSearchPaths(ctx, invoker)
 	if err != nil {
-		db_local.StopServices(false, constants.InvokerService, nil)
+		db_local.StopServices(context.Background(), false, constants.InvokerService, nil)
```
Comment (kaidaguerre) [inline cmd/service.go:195]: why background?
add comment if there is a reason

File: cmd/service.go [hunk 2]
```diff
@@ -224,7 +228,7 @@ func runServiceInForeground(invoker constants.Invoker) {
 			}
 		case <-sigIntChannel:
 			fmt.Print("\r")
-			count, err := db_local.GetCountOfConnectedClients()
+			count, err := db_local.GetCountOfConnectedClients(context.Background())
```
Comment (kaidaguerre) [inline cmd/service.go:231]: why background?

File: db/db_client/db_client.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ type DbClient struct {
 func NewDbClient(connectionString string) (*DbClient, error) {
 	utils.LogTime("db_client.NewDbClient start")
 	defer utils.LogTime("db_client.NewDbClient end")
-	db, err := establishConnection(connectionString)
+	db, err := establishConnection(context.TODO(), connectionString)
```
Comment (kaidaguerre) [inline db/db_client/db_client.go:47]: please raise an issue each time we need to add context.TODO()

File: db/db_local/execute.go [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
+	"context"
 	"database/sql"
 
 	"github.com/turbot/steampipe/constants"
 )
 
 func executeSqlAsRoot(statements ...string) ([]sql.Result, error) {
 	var results []sql.Result
-	rootClient, err := createLocalDbClient(&CreateDbOptions{Username: constants.DatabaseSuperUser})
+	rootClient, err := createLocalDbClient(context.Background(), &CreateDbOptions{Username: constants.DatabaseSuperUser})
```
Comment (kaidaguerre) [inline db/db_local/execute.go:12]: can't we pass context in to executeSqlAsRoot

File: db/db_local/local_db_client.go [hunk 5]
```diff
@@ -76,7 +76,7 @@ func (c *LocalDbClient) Close() error {
 			return err
 		}
 	}
-	ShutdownService(c.invoker)
+	ShutdownService(context.Background(), c.invoker)
```
Comment (kaidaguerre) [inline db/db_local/local_db_client.go:79]: why background?

File: interactive/run.go [hunk 6]
```diff
@@ -15,7 +17,7 @@ func RunInteractivePrompt(initChan *chan *db_common.QueryInitData) (*queryresult
 	interactiveClient, err := newInteractiveClient(initChan, resultsStreamer)
 	if err != nil {
 		utils.ShowErrorWithMessage(err, "interactive client failed to initialize")
-		db_local.ShutdownService(constants.InvokerQuery)
+		db_local.ShutdownService(context.Background(), constants.InvokerQuery)
```
Comment (kaidaguerre) [inline interactive/run.go:20]: why background

File: ociinstaller/ocidownloader.go [hunk 7]
```diff
@@ -13,18 +13,18 @@ import (
 
 type ociDownloader struct {
 	resolver remotes.Resolver
-	context  context.Context
-	Images   []*SteampipeImage
+	// context  context.Context
```
Comment (kaidaguerre) [inline ociinstaller/ocidownloader.go:16]: can we delete?

File: ociinstaller/ocidownloader.go [hunk 8]
```diff
@@ -13,18 +13,18 @@ import (
...
 // NewOciDownloader :: creats and return a ociDownloader instance
-func NewOciDownloader(ctx context.Context) *ociDownloader {
+func NewOciDownloader() *ociDownloader {
 	// oras uses containerd, which uses logrus and is set up to log
 	// warning and above.  Set to ErrrLevel to get rid of unwanted error message
 	logrus.SetLevel(logrus.ErrorLevel)
 	return &ociDownloader{
 		resolver: docker.NewResolver(docker.ResolverOptions{}),
-		context:  ctx,
+		// context:  ctx,
```
Comment (kaidaguerre) [inline ociinstaller/ocidownloader.go:27]: delete

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Comment by binaek** [comment]
@kaidaguerre  made a few changes ... can you do a quick review when you get the chance?

File: db/db_local/stop_services.go [hunk 9]
```diff
@@ -103,16 +104,16 @@ func StopServices(force bool, invoker constants.Invoker, spinner *spinner.Spinne
 	pluginManagerStopError := plugin_manager.Stop()
 
 	// stop the DB Service
-	stopResult, dbStopError := stopDBService(spinner, force)
+	stopResult, dbStopError := stopDBService(context.Background(), spinner, force)
```
Comment (kaidaguerre) [inline db/db_local/stop_services.go:107]: comment as to why background ok (or remove)

File: ociinstaller/ocidownloader.go [hunk 10]
```diff
@@ -13,18 +13,16 @@ import (
 
 type ociDownloader struct {
 	resolver remotes.Resolver
-	context  context.Context
 	Images   []*SteampipeImage
 }
 
 // NewOciDownloader :: creats and return a ociDownloader instance
```
Comment (kaidaguerre) [inline ociinstaller/ocidownloader.go:19]: update comment to new format

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: ociinstaller/ocidownloader.go (hunk 10 above)
Comment (binaek) [reply ociinstaller/ocidownloader.go]: 👍

File: ociinstaller/plugin.go [hunk 11]
```diff
@@ -13,14 +13,14 @@ import (
 )
 
 // InstallPlugin :: Install a plugin from an OCI Image
```
[the remark this answers is not in the document]
Comment (binaek) [reply ociinstaller/plugin.go]: :+1

File: db/db_local/stop_services.go (hunk 9 above)
Comment (binaek) [reply db/db_local/stop_services.go]: :+1:

File: db/db_local/start_services.go [hunk 12]
```diff
@@ -474,52 +489,36 @@ func ensurePgExtensions(databaseName string) error {
...
 // this is done during database installation, but we need to migrate current installations
-func ensureTempTablePermissions(databaseName string) error {
-	rootClient, err := createLocalDbClient(&CreateDbOptions{DatabaseName: databaseName, Username: constants.DatabaseSuperUser})
-	if err != nil {
-		return err
-	}
-	defer rootClient.Close()
-	_, err = rootClient.Exec(fmt.Sprintf("grant temporary on database %s to %s", databaseName, constants.DatabaseUser))
+func ensureTempTablePermissions(ctx context.Context, databaseName string, rootClient *sql.DB) error {
+	_, err := rootClient.Exec(fmt.Sprintf("grant temporary on database %s to %s", databaseName, constants.DatabaseUser))
```
[the remark this answers is not in the document]
Comment (binaek) [reply db/db_local/start_services.go]: :+1:

File: db/db_local/start_services.go [hunk 13]
```diff
@@ -452,19 +472,14 @@ func setupLogCollector(postgresCmd *exec.Cmd) (chan string, func(), error) {
...
 
 	errors := []error{}
-	rootClient, err := createLocalDbClient(&CreateDbOptions{DatabaseName: databaseName, Username: constants.DatabaseSuperUser})
-	if err != nil {
-		return err
-	}
-	defer rootClient.Close()
 	for _, extn := range extensions {
-		_, err = rootClient.Exec(fmt.Sprintf("create extension if not exists %s", db_common.PgEscapeName(extn)))
+		_, err := rootClient.Exec(fmt.Sprintf("create extension if not exists %s", db_common.PgEscapeName(extn)))
```
[the remark this answers is not in the document]
Comment (binaek) [reply db/db_local/start_services.go:483]: :+1:

File: connection_watcher/connection_watcher.go [hunk 14]
```diff
@@ -61,6 +62,9 @@ func (w *ConnectionWatcher) handleFileWatcherEvent(e []fsnotify.Event) {
 		}
 	}()
 
+	// this is a file system event handler and not bound to any context
+	workingCtx := context.Background()
```
[the remark this answers is not in the document]
Comment (binaek) [reply connection_watcher/connection_watcher.go]: :+1:

File: cmd/service.go [hunk 15]
```diff
@@ -239,6 +244,7 @@ func runServiceInForeground(invoker constants.Invoker) {
 			}
 			fmt.Println("Stopping Steampipe service.")
 
+			// no context to pass on. use bacckground
```
[the remark this answers is not in the document]
Comment (binaek) [reply cmd/service.go]: :+1:

File: cmd/query.go [hunk 16]
```diff
@@ -104,7 +104,12 @@ func runQueryCmd(cmd *cobra.Command, args []string) {
 	// set config to indicate whether we are running an interactive query
 	viper.Set(constants.ConfigKeyInteractive, interactiveMode)
 
-	ctx := context.Background()
+	ctx := cmd.Context()
```
[the remark this answers is not in the document]
Comment (binaek) [reply cmd/query.go:107]: i just moved the `startCancelHandler` upwards - the behavior is the same

**Pull request merged**